builtin/providers/aws/resource_aws_api_gateway_usage_plan.go

```diff
@@ -0,0 +1,498 @@
+				stage.ApiId = aws.String(v)
+			}
+
+			if v, ok := sv["stage"].(string); ok && v != "" {
+				stage.Stage = aws.String(v)
+			}
+
+			as = append(as, stage)
+		}
+
+		if len(as) > 0 {
+			params.ApiStages = as
+		}
+	}
+
+	if v, ok := d.GetOk("quota_settings"); ok {
+		settings := v.(*schema.Set).List()
+		q, ok := settings[0].(map[string]interface{})
+
+		if errors := validateApiGatewayUsagePlanQuotaSettings(q); len(errors) > 0 {
+			return fmt.Errorf("Error validating the quota settings: %v", errors)
+		}
+
+		if !ok {
+			return errors.New("At least one field is expected inside quota_settings")
+		}
+
+		qs := &apigateway.QuotaSettings{}
+
+		if sv, ok := q["limit"].(int); ok {
+			qs.Limit = aws.Int64(int64(sv))
+		}
+
+		if sv, ok := q["offset"].(int); ok {
+			qs.Offset = aws.Int64(int64(sv))
+		}
+
+		if sv, ok := q["period"].(string); ok && sv != "" {
+			qs.Period = aws.String(sv)
+		}
+
+		params.Quota = qs
+	}
+
+	if v, ok := d.GetOk("throttle_settings"); ok {
+		settings := v.(*schema.Set).List()
+		q, ok := settings[0].(map[string]interface{})
+
+		if !ok {
+			return errors.New("At least one field is expected inside throttle_settings")
+		}
+
+		ts := &apigateway.ThrottleSettings{}
+
+		if sv, ok := q["burst_limit"].(int); ok {
+			ts.BurstLimit = aws.Int64(int64(sv))
+		}
+
+		if sv, ok := q["rate_limit"].(float64); ok {
+			ts.RateLimit = aws.Float64(float64(sv))
+		}
+
+		params.Throttle = ts
+	}
+
+	up, err := conn.CreateUsagePlan(params)
+	if err != nil {
+		return fmt.Errorf("Error creating API Gateway Usage Plan: %s", err)
+	}
+
+	d.SetId(*up.Id)
+
+	// Handle case of adding the product code since not addable when
+	// creating the Usage Plan initially.
+	if v, ok := d.GetOk("product_code"); ok {
+		updateParameters := &apigateway.UpdateUsagePlanInput{
+			UsagePlanId: aws.String(d.Id()),
+			PatchOperations: []*apigateway.PatchOperation{
+				{
+					Op:    aws.String("add"),
+					Path:  aws.String("/productCode"),
+					Value: aws.String(v.(string)),
+				},
+			},
+		}
+
+		up, err = conn.UpdateUsagePlan(updateParameters)
+		if err != nil {
+			return fmt.Errorf("Error creating the API Gateway Usage Plan product code: %s", err)
+		}
+	}
+
+	return resourceAwsApiGatewayUsagePlanRead(d, meta)
+}
+
+func resourceAwsApiGatewayUsagePlanRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+	log.Printf("[DEBUG] Reading API Gateway Usage Plan: %s", d.Id())
+
+	up, err := conn.GetUsagePlan(&apigateway.GetUsagePlanInput{
+		UsagePlanId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NotFoundException" {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("name", up.Name)
+	d.Set("description", up.Description)
+	d.Set("product_code", up.ProductCode)
+
+	if up.ApiStages != nil {
+		config := flattenApiGatewayUsageApiStages(up.ApiStages)
+		d.Set("api_stages", config)
```

Comment:
I'd like us to check that the setting of this to state doesn't throw an error. i.e.

```
if err := d.Set("api_stages", flattenApiGatewayUsageApiStages(up.ApiStages)); err != nil {
			return fmt.Errorf("[DEBUG] Error setting api_stages error: %#v", err)
		}
```